.github/workflows/publish.yaml

```diff
@@ -0,0 +1,490 @@
+
+      - name: Is this running on main branch?
+        if: runner.os == 'Linux'
+        run: if [ "${{ steps.extract_branch.outputs.sha-of-current-branch}}" = "${{ steps.extract_branch.outputs.sha-of-main-branch}}" ] ; then echo "Running on main branch."; else echo "Not running on main branch."; fi
+
+      - name: Fail if attempting to publish from a branch other than main
+        if: runner.os == 'Linux' && github.event.inputs.publish_to_pypi == '1'
+        run: if [ "${{ steps.extract_branch.outputs.sha-of-current-branch}}" = "${{ steps.extract_branch.outputs.sha-of-main-branch}}" ] ; then echo "Running on main branch, can proceed to publish."; else exit 1; fi
+
+      - name: Activate virtual environment (Linux/MacOS)
+        if: runner.os == 'Linux' || runner.os == 'macOS'
+        run: . venv/bin/activate
+
+      - name: Activate virtual environment (Windows)
+        if: runner.os == 'Windows'
+        run: venv\Scripts\Activate.ps1
+
+      - name: Display Pip Version and confirm environment empty
+        run: |
+          pip -V
+          pip freeze
+      - name: Install Dev Dependencies
+        run: pip install -r requirements-dev.txt
+
+      - name: Install Main Package
+        run: pip install -e .
+      - name: Log full installed packages
+        run: pip freeze
+
+      - name: Run pre-commit hooks
+        if: matrix.IS_FASTEST_JOB == true
+        run: |
+          pre-commit install
+          pre-commit run -a
+
+      - name: Run Tests with randomly generated seed
+        if: github.event.inputs.randomlyseed == ''
+        run: pytest -xsvv --cov-report=xml --full-ci --include-slow-tests
+
+      - name: Run Tests with supplied pytest-randomly seed
+        if: github.event.inputs.randomlyseed != ''
+        run: pytest -xsvv --cov-report=xml --full-ci --include-slow-tests --randomly-seed=${{ github.event.inputs.randomlyseed }}
+
+      - name: Confirm Sphinx Docs build correctly
+        # only check Sphinx docs build on the fastest job
+        if: matrix.IS_FASTEST_JOB == true
+        run: SPHINXOPTS="-W" make --directory=docs html # the -W flag treats warnings as errors to cause build failures
+
+      - name: Run prerelease to check packaging compatibilty
+        run: prerelease --no-input
+
+      - name: Build package for release
+        run: |
+          python setup.py sdist
+          python setup.py bdist_wheel
+
+      - name: Upload build artifact
+        uses: actions/upload-artifact@v2
+        with:
+          name: dist--${{ matrix.os }}--${{ matrix.python-version }}
+          path: dist/*
+
+
+  publish_to_test_pypi:
+    needs: [test]
+    strategy:
+      fail-fast: false
+      matrix: # if publishing pure python code, then disable all except one os and python version combination. But if publishing code that uses Cython then likely all combinations will be needed
```

Comment:
when you publish code that just uses pure python, the wheels will all be generated with the same name.  So just use the latest version of linux (ubuntu 20) and the latest version of python (3.9). otherwise test.pypi and pypi will give errors saying duplicate files are attempting to be uploaded